src/osd/ReplicatedBackend.cc

```diff
@@ -154,6 +157,19 @@ bool ReplicatedBackend::can_handle_while_inactive(OpRequestRef op)
   }
 }
 
+bool ReplicatedBackend::can_handle_fast(OpRequestRef op)
+{
+  bool can_handle = false;
+  switch (op->get_req()->get_type()) {
+    case MSG_OSD_REPOPREPLY: 
```

Comment:
You *do* need a lock to ensure that these are handled in order.  It doesn't have to be the PG lock, but it does need to be *some* lock.  At the least, that lock needs to also protect the repop queue.  Please factor out any state that needs to be accessed in that path (mostly the repop queue and associated state) into a sub-object with its own lock.  That's the lock you'll need to queue the op under.